### PR TITLE
Added test for transactional JMS source

### PR DESCRIPTION
**Pull request by olukas**
hazelcast-jet-ansible part: https://github.com/hazelcast/hazelcast-jet-ansible/pull/40

### CODE REVIEW

**Comment by olukas** [comment]
FYI, I ran this test for 8hours during the night and no failure occurred during that time.

File: snapshot-jms-test/src/main/java/com/hazelcast/jet/tests/snapshot/jms/JmsSourceTest.java [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+        if (exceptions[1] != null) {
+            throw exceptions[1];
+        }
+    }
+
+    public void testInternal(JetInstance client, ILogger logger, String clusterName) throws Exception {
+        Pipeline p = Pipeline.create();
+
+        StreamSource<Long> source = Sources.jmsQueueBuilder(getConnectionFactory(brokerURL))
+                .sharedConsumer(true)
```
Comment (viliam-durina) [inline snapshot-jms-test/src/main/java/com/hazelcast/jet/tests/snapshot/jms/JmsSourceTest.java]: This is not needed for queue, but doesn't hurt.

File: snapshot-jms-test/src/main/java/com/hazelcast/jet/tests/snapshot/jms/JmsMessageProducer.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+import javax.jms.Connection;
+import javax.jms.MessageProducer;
+import javax.jms.Session;
+import javax.jms.TextMessage;
+
+import static com.hazelcast.jet.impl.util.Util.uncheckRun;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+class JmsMessageProducer {
```
Comment (gurbuzali) [inline snapshot-jms-test/src/main/java/com/hazelcast/jet/tests/snapshot/jms/JmsMessageProducer.java]: this class looks like the same as the `jms-test` producer. I think we can depend/shade it into the test jar like we do with the `soak-tests-common`

File: snapshot-jms-test/src/main/java/com/hazelcast/jet/tests/snapshot/jms/JmsSourceTest.java [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+            if (expectedTotalCount == actualTotalCount) {
+                return;
+            }
+            SECONDS.sleep(1);
+        }
+        long actualTotalCount = latestCounterMap.get(clusterName);
+        assertEquals(expectedTotalCount, actualTotalCount);
+    }
+
+    static class JmsFactorySupplier implements Serializable {
```
Comment (viliam-durina) [inline snapshot-jms-test/src/main/java/com/hazelcast/jet/tests/snapshot/jms/JmsSourceTest.java]: Probably doesn't need to be serializable.

File: snapshot-jms-test/src/main/java/com/hazelcast/jet/tests/snapshot/jms/VerificationProcessor.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    private static final int QUEUE_SIZE_LIMIT = 5_000;
+    private static final int PRINT_LOG_ITEMS = 5_000;
+
+    private final String name;
+
+    private boolean processed;
+    private long counter;
+    private final PriorityQueue<Long> queue = new PriorityQueue<>();
+    private ILogger logger;
+    private Map<String, Long> map;
```
Comment (viliam-durina) [inline snapshot-jms-test/src/main/java/com/hazelcast/jet/tests/snapshot/jms/VerificationProcessor.java]: Would be nice to have this as `IMap`, it's a lot different thing than just `Map`. Took me a while to notice that it's an imap actually.

File: snapshot-jms-test/src/main/java/com/hazelcast/jet/tests/snapshot/jms/JmsMessageProducer.java (hunk 2 above)
Comment (olukas) [reply snapshot-jms-test/src/main/java/com/hazelcast/jet/tests/snapshot/jms/JmsMessageProducer.java]: You're right, they are almost the same. I will create an issue for that and fix it after release. Is it ok?

Comment (gurbuzali) [reply snapshot-jms-test/src/main/java/com/hazelcast/jet/tests/snapshot/jms/JmsMessageProducer.java]: 👍

**Comment by olukas** [comment]
Thanks for review guys. I addressed comments and I'm running another 1hour test - if it will pass I merge this PR.

**Pull request merged**